src/mlpack/methods/reinforcement_learning/environment/ftn.hpp

```diff
@@ -0,0 +1,300 @@
+          "being taken.";
+      return true;
+    }
+    else if (state.Row() == fruitTree.Depth())
+    {
+      Log::Info << "Episode terminated due to reaching leaf node.";
+      return true;
+    }
+    return false;
+  }
+  
+  //! Get the number of steps performed.
+  size_t StepsPerformed() const { return stepsPerformed; }
+
+  //! Get the maximum number of steps allowed.
+  size_t MaxSteps() const { return maxSteps; }
+  //! Set the maximum number of steps allowed.
+  size_t& MaxSteps() { return maxSteps; }
+
+  //! The reward vector consists of {Protein, Carbs, Fats, Vitamins, Minerals, Water}
+  //! A total of 6 rewards.
+  static constexpr size_t rewardSize = 6;
+
+ private:
+
+  /*
+   *  A Fruit Tree is a full binary tree. Each node of this tree represents a state.
+   *  Non-leaf nodes yield a null vector R^{6} reward. Leaf nodes have a pre-defined
+   *  set of reward vectors such that they lie on a Convex Convergence Set (CCS).
+   */
+  class FruitTree
+  {
+    public:
+      FruitTree(const size_t depth) : depth(depth)
+      { 
+        if (std::find(validDepths.begin(), validDepths.end(), depth) == validDepths.end())
+        {
+          throw std::logic_error("FruitTree()::FruitTree: Invalid depth value: " + std::to_string(depth) + " provided. "
+                                 "Only depth values of: 5, 6, 7 are allowed.");
+        }
+
+        arma::mat branches = arma::zeros((rewardSize, (size_t) std::pow(2, depth - 1)));
+        tree = arma::join_rows(branches, Fruits());
+      }
+
+      // Extract array index from {(row, column)} representation.
+      size_t GetIndex(const State& state) const
+      {
+        return static_cast<size_t>(std::pow(2, state.Row() - 1) + state.Column());
+      }
+
+      // Yield reward from the current node.
+      arma::vec GetReward(const State& state) const
+      {
+        return tree.col(GetIndex(state));
+      }
+
+      //! Retreive the reward vectors of the leaf nodes.
+      arma::mat Fruits() const { return ConvexSetMap.at(depth); }
+
+      //! Retreive the maximum depth of the FTN tree.
+      size_t Depth() const { return depth; }
+    
+    private:
+      //! Maximum depth of the tree.
+      size_t depth;
+
+      //! Matrix representation of the internal tree.
+      arma::mat tree;
+
+      //! Valid depth values for the tree.
+      const std::array<size_t, 3> validDepths {5, 6, 7};
+
+      //! Maps depth value to the CCS.
+      const std::unordered_map<size_t, arma::mat> ConvexSetMap =
+      {
+        {
+          '5', arma::mat(
+              {
+                {3.67917966,7.49190652,3.14072363,0.03085158,4.02109647,2.96104264,3.95849765,0.74185053,4.28506840,5.93102382,1.05416268,0.79064992,4.35145246,2.44514113,6.66924742,4.45750971,1.70141153,1.61886242,2.36047279,0.52087376,4.85800708,3.51780254,0.92487495,0.91807423,1.00725144,1.89889027,4.57737545,1.23346934,1.51947586,3.81593770,5.85626898,4.24912001}, 
```

Comment:
Another nitpick, can we parse these values from a file instead of storing a bunch of depth values in the source itself?

Going through the source, it appears as an wall of text! 🕵🏽

Reply:
Makes sense to me. @zoq  Your thoughts?

Reply:
Maybe outsource it in another hpp file? So it#s still backed in but kinda outsourced?

Reply:
Okay, I should move the FruitTree class to `.hpp`, right?

Reply:
I think what Marcus meant was to move the numbers (solely) to another header file and then use it in `ftn.hpp`.

Reply:
Yeah I got that, but I was thinking we might as well put the entire `FruitTree` class in a separate `.hpp` or is it a bad idea??

Reply:
Right I was thinking we could create another header file just for the parameters e.g. `ftn_param.hpp` or something, in that file we could use `static constexpr` to define the parameters in the correct namespace.

Isn't the `FruitTree` class already a separate file? This extra `hpp` for the parameters follows @say4n idea to outsource the huge parameter lists to make the code more readable.

Reply:
@say4n @zoq  Aight, just updated, see if it checks out??

Reply:
LGTM! :)